### PR TITLE
CPA LockoutPanel UI

### PR DESCRIPTION
**Pull request by wilkie**
This adds `LockoutPanel` and the associated route, images, haml page, and JS initialization in our pipeline.

There is no behavior, just the UI, as per our ticket. Gluing this UI and the API/data model together is a next step.

But I did go ahead and add a route for the page for convenience.

Added each respective thingy to the `sessions` namespace... not sure if that's a good place. Felt right and better than creating something new or have it float around with everything else.

### Behaviors of the UI:

* Not currently masked behind the CPA experiment flag... not necessary: not in any common path or updating any already used UI.
* `/lockout` route is a placeholder... any failed authentication on parental permission should just redirect there.
* Information is passed by adding data attributes on the containing div currently. Easy to change.
* Submit button is disabled until a valid email is typed in.
* Resend button (styled as a link) replaces the email with the pending email and submits.
* Delete by date is rendered as UTC so there are no strange timezone shenanigans when rendering it. Let's be sure to generate a UTC delete by date.
* Last request date is rendered in the current locale / timezone. We should still store it as UTC.

### New Strings for i18n:

`sessionLockoutNewAccountHeader`: "Just one more thing!"
`sessionLockoutPendingHeader`: "Thanks! We've contacted your parent."
`sessionLockoutPrompt`: "We need your parent to approve your account before you can get started. Please supply us with your parent’s email address so they can grant you permission."
`sessionLockoutPendingPrompt`: "We sent an email to {pendingEmail}. Didn’t receive anything? Update your parent’s email below or send another request."
`sessionLockoutNote`: "Note: Your account will be deleted if we do not receive your parent’s permission by {deleteDate}."
`sessionLockoutParentEmailField`: "Parent Email:"
`sessionLockoutParentStatusField`: "Parent Permission:"
`sessionLockoutSubmit`: "Send permission request"
`sessionLockoutStatusPending`: "Pending"
`sessionLockoutStatusNotSubmitted`: "Not Submitted"
`sessionLockoutResendEmail`: "Resend permission email"
`sessionLockoutUpdateSubmit`: "Update and send"
`signOutButton`: "Sign out",

## Links

- jira ticket: [P20-61](https://codedotorg.atlassian.net/browse/P20-61)

## Testing story

Manual testing of the basic boilorplate behaviors. Then, contributing one Storybook with two templates:

### New Accounts

image

### When there is a pending email request:

image

### New Accounts (right-to-left)

image

### Pending email (right-to-left)

image

## Deployment strategy

No experiment flag. We can just deploy the UI without the glue since nobody will know to look at it yet. :)

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

**Comment by wilkie** [comment]
Amended quickly to fix a commented out line... I was debugging. Probably got that in before anybody saw. :)

File: apps/i18n/common/en_us.json
Comment (carl-codeorg) [inline apps/i18n/common/en_us.json]: Is a provision in the law for a legal guardian (but not a parent) to provide permission? If so do we need language like "parent or guardian"?

File: apps/i18n/common/en_us.json
Comment (elf-code) [inline apps/i18n/common/en_us.json:2131]: Is there an existing string for this? (I see it in English and Spanish so there's something that gets translated that means Sign Out)

File: apps/src/templates/sessions/LockoutPanel.jsx
Comment (elf-code) [inline apps/src/templates/sessions/LockoutPanel.jsx]: Can we make a new React component a functional component instead of a class component? I'm happy to pair with you to shift around the syntax or twist it around myself if that's easier.

**Review by elf-code — COMMENTED** [review]
Is there a way to directly view the Storybook as part of testing? Can you include the path/screenshot to that (for learning purposes)?

**Comment by elf-code** [comment]
Also love the PR! Cool to see a new route and UI!

File: apps/src/templates/sessions/LockoutPanel.jsx
Comment (daynew) [inline apps/src/templates/sessions/LockoutPanel.jsx]: My understanding is that it's better to use [Flexbox](https://css-tricks.com/snippets/css/a-guide-to-flexbox/) for layout stuff instead of trying to use tables or table styling. Did you already try using it?

**Review by daynew — CHANGES_REQUESTED** [review]

File: apps/i18n/common/en_us.json
Comment (daynew) [reply apps/i18n/common/en_us.json:2131]: I don't think there is a "Sign Out" in this file already.

File: apps/src/templates/sessions/LockoutPanel.jsx
Comment (wilkie) [reply apps/src/templates/sessions/LockoutPanel.jsx]: That's kinda true. I could try to use a full width flex box for that section and grow the field side. I think if I establish the flex basis, it won't squish the text. Hmm. Yes, I think that would do the same thing.

File: apps/i18n/common/en_us.json
Comment (wilkie) [reply apps/i18n/common/en_us.json:2131]: No, there is no established string for "Sign out" available to the front-end. 🥲

File: apps/src/templates/sessions/LockoutPanel.jsx
Comment (wilkie) [reply apps/src/templates/sessions/LockoutPanel.jsx]: Do you have an example of a functional component that is similar to our case here somewhere in our codebase to use to conform to convention?

File: apps/i18n/common/en_us.json
Comment (wilkie) [reply apps/i18n/common/en_us.json]: That is a good question. I noticed this language already used on our existing parent email process, so it seems like a good idea to ask to see if we should just conform to that language.

File: apps/src/templates/sessions/LockoutPanel.jsx
Comment (wilkie) [reply apps/src/templates/sessions/LockoutPanel.jsx]: I'd particularly like an example of a component with form interactivity... that's the challenge I see in re-wrapping it in the functional style.

**Comment by wilkie** [comment]
> Is there a way to directly view the Storybook as part of testing? Can you include the path/screenshot to that (for learning purposes)?

The screenshots of the storybooks are more or less how they are depicted in the PR description. Our storybook view does not have root application-level styling, so the headers look wrong (as they do on all other storybook components), but otherwise look the same as the site.

You view the storybooks by pulling down my branch and then running the storybook web server and visiting it (localhost, port 9001, I think):

```
cd apps
yarn storybook
```

I do not believe we have a staging storybook server set up at all? That'd be nice.

image

(Ignore the broken image... it works... I'm just too lazy to restart my web server)

image

File: apps/src/templates/sessions/LockoutPanel.jsx
Comment (wilkie) [reply apps/src/templates/sessions/LockoutPanel.jsx]: Gosh expanding `<input>` tags to the full width of grown divs in a flexbox is silly annoying, but I have done it... will test it a little and push it up.

**Comment by wilkie** [comment]
I added screenshots of the right-to-left rendering of the widget. We do not have a convention for creating storybooks for right-to-left... nor do we have a strategy of rendering components right-to-left in the storybook view. Hmm.

File: apps/src/templates/sessions/LockoutPanel.jsx
Comment (wilkie) [reply apps/src/templates/sessions/LockoutPanel.jsx]: Done.

File: apps/src/templates/sessions/LockoutPanel.jsx
Comment (elf-code) [reply apps/src/templates/sessions/LockoutPanel.jsx]: Here's an example in our codebase (I think it has a lot of interactions) -- https://github.com/code-dot-org/code-dot-org/blob/staging/apps/src/music/views/ChordPanel.tsx

Hooks are weird (I think useState is the only semi-intuitive one, but the example component has a couple different hooks in it). But the React.dev notes on hooks are good - https://react.dev/reference/react

Comment (wilkie) [reply apps/src/templates/sessions/LockoutPanel.jsx]: This is helpful! Good to see the conventions to follow... I already did the work last night... pretty straightforward. Much prefer the "functional" style.

Comment (wilkie) [reply apps/src/templates/sessions/LockoutPanel.jsx]: Done. Ish. Do I need a `useCallback` if there's no state being used? It's just reacting to input events. It's always the same function.

Global redux state can't be wrapped very effectively, I don't think... but the things being pulled from that are the `isRTL` bool which can only be changed via changing the locale which is always a page refresh anyway.

Comment (elf-code) [reply apps/src/templates/sessions/LockoutPanel.jsx]: I don't think you need a `useCallback` and yes, I agree on the redux store and `isRTL`

**Review by elf-code — APPROVED** [review]
LGTM! Also I'm assuming you double-checked that the functional changes don't change the visuals.

**Comment by wilkie** [comment]
Indeed. In fact, the PR description has been updated with new visual screens which include RTL rendering. I'll verify that the storybooks render later today before merging, once I get Dayne's approval on their requested changes.

File: apps/src/templates/sessions/LockoutPanel.jsx
Comment (daynew) [inline apps/src/templates/sessions/LockoutPanel.jsx:64]: alt text for both images

File: apps/src/templates/sessions/LockoutPanel.jsx
Comment (daynew) [inline apps/src/templates/sessions/LockoutPanel.jsx:56]: I wonder if the `EnhancedSafeMarkdown` component would work? I think that is how we render markdown strings in other React components. I think your solution works as well though.

File: dashboard/config/routes.rb
Comment (daynew) [inline dashboard/config/routes.rb:200]: What if a student visits this URL and they have permission? I'm thinking we should redirect them to the /home page, but that work should be done in the global redirect task. I will note it there.

File: apps/src/templates/sessions/LockoutPanel.jsx
Comment (daynew) [inline apps/src/templates/sessions/LockoutPanel.jsx]: Oh cool, I didn't know you had to use `htmlFor` instead of `for` in JSX. That makes sense. Thanks for looking this up :)

File: apps/src/templates/sessions/LockoutPanel.jsx
Comment (daynew) [inline apps/src/templates/sessions/LockoutPanel.jsx]: Maybe a nit, but should we use flexbox to avoid using `float` and `isRTL` customization? Would the flexbox `justify-content: space-between` achieve the same spacing between the "Sign out" and "Send permission request" buttons?

**Review by daynew — APPROVED** [review]
Looks great! Just some minor comments.

File: apps/src/templates/sessions/LockoutPanel.jsx
Comment (wilkie) [reply apps/src/templates/sessions/LockoutPanel.jsx]: This is an amazing idea.

**Comment by wilkie** [comment]
Will be addressing Dayne's comments and adding the new text for Parent/Guardian for the next pass.

File: apps/i18n/common/en_us.json
Comment (wilkie) [reply apps/i18n/common/en_us.json]: We are updating the text! Pushing that up with the next commit.

**Review by carl-codeorg — APPROVED** [review]

File: apps/src/templates/sessions/LockoutPanel.jsx
Comment (wilkie) [reply apps/src/templates/sessions/LockoutPanel.jsx:64]: I've added them and the strings will be localized.

File: apps/src/templates/sessions/LockoutPanel.jsx
Comment (wilkie) [reply apps/src/templates/sessions/LockoutPanel.jsx:56]: I'm gonna leave it as is, then. I don't think it is worth potentially overcomplicating the string and worry about what happens if markdown ends up in the translated strings. Pity there isn't a good way of just doing string interpolation with localized strings and injecting simple markdown or html without needlessly going with markdown 100%.

File: apps/src/templates/sessions/LockoutPanel.jsx
Comment (wilkie) [reply apps/src/templates/sessions/LockoutPanel.jsx]: That worked really well and now I want to revisit the past 10 years of my life and find all the places where I should have done this. Thank you!

**Comment by wilkie** [comment]
Ok. Added the new "parent / guardian" strings. I think it covers all of the cases. Removed all "float"ed content and replaced with some cool flexbox ideas that will work in both LTR and RTL circumstances. I have looked at the visuals and confirm that they still match the ones in the description. No visual changes. Alt text applied to both header images to describe the little penguin :)

File: dashboard/app/views/sessions/lockout.html.haml
Comment (artem-vavilov) [inline dashboard/app/views/sessions/lockout.html.haml]: Just a small recommendation
```suggestion
#lockout-container{data: {request_date: 6.days.ago.iso8601, delete_date: 2.days.since.iso8601, pending_email: '', api_url: '/permissions'}}
```

**Pull request merged**
